### ISSUE TITLE
Question: Moya not respecting SessionManager configuration for requests?

### ISSUE THREAD

**Issue by pschneider** [Moya/Moya#700]
Hi all,

I'm currently trying to implement a cache solution with a custom URLCache (and modify some cache methods like in this example (http://stackoverflow.com/a/19847811), Moya/Alamofire and therefor I'm using a custom SessionManager with my provider which configures the URLCache to use.

Especially this part is interesting:

``` swift
// ...
URLCache.shared = myCustomURLCache
sessionConfig.urlCache = URLCache.shared
sessionConfig.requestCachePolicy = .returnCacheDataElseLoad
// ...
```

I pass the final manager to my `MoyaProvider`, but it seems, that the send request does not have the correct `RequestCachePolicy` set.

Now if I put some output in 
`func sendRequest(_ target: Target, request: URLRequest, queue: DispatchQueue?, progress: Moya.ProgressBlock?, completion: @escaping Moya.Completion) -> CancellableToken {`

I'm seeing this:

``` swift
print(manager.session.configuration.requestCachePolicy.rawValue) // correctly set
print(request.cachePolicy.rawValue) // wrong (always `useProtocol...`)
let alamoRequest = manager.request(request as URLRequestConvertible)
print(alamoRequest.request?.cachePolicy.rawValue) // wrong
```

I don't know if this is intended or maybe a bug in Alamofire as Moya is using the correctly configured `manager` to build the `alamoRequest`.

Is this behavior intended in any way?

**Comment by DanteCore** [comment]
Also have this problem. I tried to set custom SessionManager with RequestRetrier that was introduced in Alamofire 4, but implementation of RequestRetrier protocol not invokes

**Comment by jeryRazakarison** [comment]
In the case of the requestRetrier, that is because the requestRetrier delegate method is **only** called when there is an **Alamofire error** in the response. (you can see it here https://github.com/Alamofire/Alamofire/blob/master/Source/SessionDelegate.swift#L434)
When you directly use Alamofire, this error would appear if you call `validate()` on a DataRequest.

BUT when using Moya, you can't call `validate()` on a DataRequest (since it is hidden), and thus Alamofire do not consider there is an error, and the requestRetrier `should` process is not launched.

You can test that your requestRetrier would still work with Moya by calling validate() on the DataRequest here:
https://github.com/Moya/Moya/blob/master/Source/Moya%2BInternal.swift#L187
would be
`let alamoRequest = manager.request(request as URLRequestConvertible).validate()`
Using this **hot fix**, your requestAdapter and requestRetrier works as before with Alamofire.

But the real improvement would be for Moya to allow us in the switch/case configuration (in the `TargetType` protocol) to pass true or false, if we want or not the Alamofire `validate()` appended.

**Comment by BasThomas** [comment]
This seems like a great addition! Would you be able to submit a PR implementing this, @jeryRazakarison?

**Comment by jeryRazakarison** [comment]
Yes, I will do that tomorrow. I tested it and it works well :)

**Comment by ashfurrow** [comment]
Closed by #767, thanks @jeryRazakarison!
